p2p/udp.go

```diff
@@ -0,0 +1,226 @@
+package p2p
+
+import (
+	"errors"
+	"fmt"
+	"net"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/spacemeshos/go-spacemesh/log"
+	inet "github.com/spacemeshos/go-spacemesh/p2p/net"
+	"github.com/spacemeshos/go-spacemesh/p2p/node"
+	"github.com/spacemeshos/go-spacemesh/p2p/p2pcrypto"
+	"github.com/spacemeshos/go-spacemesh/p2p/pb"
+	"github.com/spacemeshos/go-spacemesh/p2p/service"
+)
+
+const maxMessageSize = 2048
+
+// todo : calculate real udp max message size
+
+// Lookuper is a service used to lookup for nodes we know already
+type Lookuper interface {
+	InternalLookup(key node.DhtID) []node.Node
+}
+
+type udpNetwork interface {
+	Start() error
+	Shutdown()
+
+	IncomingMessages() chan inet.UDPMessageEvent
+	Send(to node.Node, data []byte) error
+}
+
+// UDPMux is a server for receiving and sending udp messages. through protocols.
+type UDPMux struct {
+	local    *node.LocalNode
+	lookuper Lookuper
+	network  udpNetwork
+	messages map[string]chan service.DirectMessage
+	shutdown chan struct{}
+	logger   log.Log
+}
+
+// NewUDPMux creates a new udp protocol server
+func NewUDPMux(localNode *node.LocalNode, lookuper Lookuper, udpNet udpNetwork, logger log.Log) *UDPMux {
+	return &UDPMux{
+		localNode,
+		lookuper,
+		udpNet,
+		make(map[string]chan service.DirectMessage),
+		make(chan struct{}, 1),
+		logger,
+	}
+}
+
+// Start starts the UDPMux
+func (mux *UDPMux) Start() error {
+	err := mux.network.Start()
+	if err != nil {
+		return err
+	}
+	go mux.listenToNetworkMessage()
+	return nil
+}
+
+// Shutdown closes the server
+func (mux *UDPMux) Shutdown() {
+	close(mux.shutdown)
+	mux.network.Shutdown()
+}
+
+func (mux *UDPMux) listenToNetworkMessage() {
+	msgChan := mux.network.IncomingMessages()
+	for {
+		select {
+		case msg, ok := <-msgChan:
```

Comment:
this means you are always reading from this channel, when the channel closes, doesn't it return nil to signal it?

Reply:
do you mean it keeps reading after the channel is closed ? either way, it will only happen once. as we return when that's the case.